src/main/java/address/sync/CloudSimulator.java

```diff
@@ -1,129 +1,670 @@
+    }
+
+    private void modifyCloudTagListBasedOnChance(List<CloudTag> cloudTagList) {
+        cloudTagList.stream()
+                .forEach(cloudTag -> modifyCloudTagBasedOnChance(cloudTag));
+    }
+
+    private void addCloudPersonsBasedOnChance(List<CloudPerson> personList) {
+        for (int i = 0; i < MAX_NUM_PERSONS_TO_ADD; i++) {
+            if (RANDOM_GENERATOR.nextDouble() <= ADD_PERSON_PROBABILITY) {
+                CloudPerson person = new CloudPerson(java.util.UUID.randomUUID().toString(),
+                                                     java.util.UUID.randomUUID().toString());
+                System.out.println("Cloud simulator: adding " + person);
+                personList.add(person);
             }
-            modifiedData.add(person);
         }
+    }
+
+    private void addCloudTagsBasedOnChance(List<CloudTag> tagList) {
+        for (int i = 0; i < MAX_NUM_PERSONS_TO_ADD; i++) {
+            if (RANDOM_GENERATOR.nextDouble() <= ADD_TAG_PROBABILITY) {
+                CloudTag tag = new CloudTag(java.util.UUID.randomUUID().toString());
+                System.out.println("Cloud simulator: adding tag '" + tag + "'");
+                tagList.add(tag);
+            }
+        }
+    }
+
+    private void modifyCloudPersonBasedOnChance(CloudPerson cloudPerson) {
+        if (RANDOM_GENERATOR.nextDouble() > MODIFY_PERSON_PROBABILITY) return;
+        System.out.println("Cloud simulator: modifying person '" + cloudPerson + "'");
+        cloudPerson.setCity(java.util.UUID.randomUUID().toString());
+        cloudPerson.setStreet(java.util.UUID.randomUUID().toString());
+        cloudPerson.setPostalCode(String.valueOf(RANDOM_GENERATOR.nextInt(999999)));
+    }
+
+    private void modifyCloudTagBasedOnChance(CloudTag cloudTag) {
+        if (RANDOM_GENERATOR.nextDouble() > MODIFY_TAG_PROBABILITY) return;
+        System.out.println("Cloud simulator: modifying tag '" + cloudTag + "'");
+        cloudTag.setName(UUID.randomUUID().toString());
+    }
+
+    private void delayRandomAmount() {
+        long delayAmount = RANDOM_GENERATOR.nextInt(DELAY_RANGE) + MIN_DELAY_IN_SEC;
+        try {
+            TimeUnit.SECONDS.sleep(delayAmount);
+        } catch (InterruptedException e) {
+            System.out.println("Error occurred while adding cloud response delay.");
+        }
+    }
+
+    private ByteArrayInputStream convertToInputStream(Object object) {
+        try {
+            return new ByteArrayInputStream(JsonUtil.toJsonString(object).getBytes());
+        } catch (JsonProcessingException e) {
+            e.printStackTrace();
+            return null;
+        }
+    }
+
+    private int getNumberOfRequestsRequired(int dataSize, int resourcesPerPage) {
+        return (int) Math.ceil((double) dataSize / resourcesPerPage);
+    }
+
+    private void deletePersonFromData(List<CloudPerson> personList, String firstName, String lastName)
```

Comment:
This does not delete the person from the list directly. It simply marks the person as `deleted`. However, the local model does not account for this (yet) & the `CloudService` simply ignores the deleted field.

Note that removing the person from the list directly may also cause problems, such as not being able to determine what has been deleted on the cloud without downloading the full list of data.